casper/src/test/resources/EitherTest.rho

```diff
@@ -25,31 +48,174 @@ in {
+        ch1!(Nil) | ch2!("Not Nil") |
+        @Either!("fromNillableError <-", *ch1, "If not Nil", *ch1ret) |
+        @Either!("fromNillableError <-", *ch2, "If not Nil", *ch2ret) |
+        rhoSpec!("assertMany",
+          [
+            (((true, Nil), "== <-", *ch1ret), "Uses Nil when Nil"),
+            (((false, "If not Nil"), "== <-", *ch2ret), "Uses given error when not Nil")
+          ], *ackCh)
+      }
+    } |
+
+    contract test_product_r_arrow(rhoSpec, _, ackCh) = {
+      new a1, b1, ret1, a2, b2, ret2 in {
+        a1!((false, "Error")) | b1!((true, "Value")) |
+        a2!((true, "A value")) | b2!((true, "B value")) |
+        @Either!("productR <-", *a1, *b1, *ret1) |
+        @Either!("productR <-", *a2, *b2, *ret2) |
+        rhoSpec!("assertMany",
+          [
+            (((false, "Error"), "== <-", *ret1), "Uses error message when Left"),
+            (((true, "B value"), "== <-", *ret2), "Uses B value when Right")
+          ], *ackCh)
+      }
+    } |
+
+    contract test_map2_arrow(rhoSpec, _, ackCh) = {
+      new ch1a, ch2a, reta,
+          ch1b, ch2b, retb,
+          ch1c, ch2c, retc,
+          f in {
+        ch1a!((false, 1)) | ch2a!((true, 2)) |
+        ch1b!((true, 3)) | ch2b!((false, 4)) |
+        ch1c!((true, 5)) | ch2c!((true, 6)) |
+        @Either!("map2 <-", *ch1a, *ch2a, for (@x, @y, r <- f) { r!(x + y) }, *reta) |
+        @Either!("map2 <-", *ch1b, *ch2b, for (@x, @y, r <- f) { r!(x + y) }, *retb) |
+        @Either!("map2 <-", *ch1c, *ch2c, for (@x, @y, r <- f) { r!(x + y) }, *retc) |
+        rhoSpec!("assertMany",
+          [
+            (((false, 1), "== <-", *reta), "If a is Left, return a"),
+            (((false, 4), "== <-", *retb), "If a is Right and b is Left, return b"),
+            (((true, 11), "== <-", *retc), "If a and b are Right, apply f")
+          ], *ackCh)
+      }
+    } |
+
+    contract test_map2_clean_arrow(rhoSpec, _, ackCh) = {
+      new ch1a, ch2a, reta,
+          ch1b, ch2b, retb,
+          ch1c, ch2c, retc,
+          f in {
+        ch1a!((false, 1)) | ch2a!((true, 2)) |
+        ch1b!((true, 3)) | ch2b!((false, 4)) |
+        ch1c!((true, 5)) | ch2c!((true, 6)) |
+        @Either!("map2Clean <-", *ch1a, *ch2a, for (@state, @x, @y, r <- f) { r!(x + y + state) }, 7, *reta) |
+        @Either!("map2Clean <-", *ch1b, *ch2b, for (@state, @x, @y, r <- f) { r!(x + y + state) }, 7, *retb) |
+        @Either!("map2Clean <-", *ch1c, *ch2c, for (@state, @x, @y, r <- f) { r!(x + y + state) }, 7, *retc) |
         rhoSpec!("assertMany",
           [
-            (((false, "message"), "== <-", *ch1), "Left values are not affected by map"),
-            (((true, 6), "== <-", *ch2), "Right values are changed by map")
+            (((false, 1), "== <-", *reta), "If a is Left, return a"),
+            (((false, 4), "== <-", *retb), "If a is Right and b is Left, return b"),
+            (((true, 18), "== <-", *retc), "If a and b are Right, apply f")
```

Comment:
I was wondering where 18 came from. Maybe extracting `7` to `chInitialState!!(7)` and then using that channel in test call would help with noticing that?

Reply:
Changed 18 to 5+6+7.